com.hazelcast.cp.internal.CPMemberAddRemoveTest.when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot [HZ-1398]

_master_ (commit 504ac9f2f3e2a561087708436085d9fb83b3ace5)

Failed on IbmJDK8 in FIPS mode: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-IbmJDK8-fips-nightly/420/testReport/com.hazelcast.cp.internal/CPMemberAddRemoveTest/when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot/

<details><summary>Stacktrace:</summary>

```
java.lang.AssertionError: expected:<[CPMember{uuid=d594f66e-5397-4d70-8030-741f2c444a27, address=[127.0.0.1]:5702}, CPMember{uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5, address=[127.0.0.1]:5703}, CPMember{uuid=e6275a2d-ffd3-4320-9381-a1dfecbd8adf, address=[127.0.0.1]:5704}]> but was:<[CPMember{uuid=d594f66e-5397-4d70-8030-741f2c444a27, address=[127.0.0.1]:5702}, CPMember{uuid=191f010e-ad74-4e4e-b7eb-6aab64c10df5, address=[127.0.0.1]:5701}, CPMember{uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5, address=[127.0.0.1]:5703}]>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at com.hazelcast.cp.internal.CPMemberAddRemoveTest.lambda$when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot$18(CPMemberAddRemoveTest.java:812)
	at com.hazelcast.cp.internal.CPMemberAddRemoveTest$$Lambda$11339/00000000244441E0.run(Unknown Source)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1226)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1328)
	at com.hazelcast.cp.internal.CPMemberAddRemoveTest.when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot(CPMemberAddRemoveTest.java:809)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:822)
```

</details>

<details><summary>Standard output:</summary>

```
05:34:21,382  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [RaftGroupMembershipManager] hz.serene_spence.cached.thread-11 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Current leadership claims:
	CPMember{uuid=d594f66e-5397-4d70-8030-741f2c444a27, address=[127.0.0.1]:5702} has 0,
	CPMember{uuid=e6275a2d-ffd3-4320-9381-a1dfecbd8adf, address=[127.0.0.1]:5704} has 0,
	CPMember{uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5, address=[127.0.0.1]:5703} has 0 leaderships.
05:34:21,382  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [RaftGroupMembershipManager] hz.serene_spence.cached.thread-11 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] CPGroup leadership balance is fine, cannot rebalance further...
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [LifecycleService] Thread-13485 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5703 is SHUTTING_DOWN
05:34:32,591  WARN |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Terminating forcefully...
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Shutting down connection manager...
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MockServer] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5703, uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5702, uuid=d594f66e-5397-4d70-8030-741f2c444a27], remoteEndpoint=[address=[127.0.0.1]:5703, uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5], alive=false}
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MockServer] Thread-13485 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5702, uuid=d594f66e-5397-4d70-8030-741f2c444a27], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5703, uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5], remoteEndpoint=[address=[127.0.0.1]:5702, uuid=d594f66e-5397-4d70-8030-741f2c444a27], alive=false}
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MockServer] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5703, uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5704, uuid=8c19428a-bf0a-4e65-bb3b-cd768e49d1c2], remoteEndpoint=[address=[127.0.0.1]:5703, uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5], alive=false}
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MockServer] Thread-13485 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5704, uuid=8c19428a-bf0a-4e65-bb3b-cd768e49d1c2], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5703, uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5], remoteEndpoint=[address=[127.0.0.1]:5704, uuid=8c19428a-bf0a-4e65-bb3b-cd768e49d1c2], alive=false}
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MembershipManager] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Removing Member [127.0.0.1]:5703 - 2311a9e1-3881-4749-8314-1934e8cf0aa5
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [PartitionStateManager] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Storing snapshot of partition assignments while removing UUID 2311a9e1-3881-4749-8314-1934e8cf0aa5
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [PartitionStateManager] hz.recursing_bardeen.priority-generic-operation.thread-0 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Storing snapshot of partition assignments while removing UUID 2311a9e1-3881-4749-8314-1934e8cf0aa5
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [ClusterService] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] 

Members {size:2, ver:6} [
	Member [127.0.0.1]:5702 - d594f66e-5397-4d70-8030-741f2c444a27 this
	Member [127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2
]

05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [TransactionManagerService] hz.pensive_spence.cached.thread-7 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5703, UUID: 2311a9e1-3881-4749-8314-1934e8cf0aa5
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Shutting down node engine...
05:34:32,591  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [ClusterService] hz.recursing_bardeen.priority-generic-operation.thread-0 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] 

Members {size:2, ver:6} [
	Member [127.0.0.1]:5702 - d594f66e-5397-4d70-8030-741f2c444a27
	Member [127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2 this
]

05:34:32,591  WARN |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [RaftService] hz.pensive_spence.cached.thread-7 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] CPMember{uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5, address=[127.0.0.1]:5703} is not present in the cluster. It will be auto-removed after 14400 seconds.
05:34:32,592  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [TransactionManagerService] hz.recursing_bardeen.cached.thread-6 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5703, UUID: 2311a9e1-3881-4749-8314-1934e8cf0aa5
05:34:32,594  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [NodeExtension] Thread-13485 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Destroying node NodeExtension.
05:34:32,594  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] Hazelcast Shutdown is completed in 3 ms.
05:34:32,594  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [LifecycleService] Thread-13485 - [127.0.0.1]:5703 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5703 is SHUTDOWN
05:34:32,594  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [LifecycleService] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5702 is SHUTTING_DOWN
05:34:32,594  WARN |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Terminating forcefully...
05:34:32,594  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Shutting down connection manager...
05:34:32,594  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MockServer] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5702, uuid=d594f66e-5397-4d70-8030-741f2c444a27], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5704, uuid=8c19428a-bf0a-4e65-bb3b-cd768e49d1c2], remoteEndpoint=[address=[127.0.0.1]:5702, uuid=d594f66e-5397-4d70-8030-741f2c444a27], alive=false}
05:34:32,594  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MockServer] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:5704, uuid=8c19428a-bf0a-4e65-bb3b-cd768e49d1c2], connection: MockConnection{localEndpoint=[address=[127.0.0.1]:5702, uuid=d594f66e-5397-4d70-8030-741f2c444a27], remoteEndpoint=[address=[127.0.0.1]:5704, uuid=8c19428a-bf0a-4e65-bb3b-cd768e49d1c2], alive=false}
05:34:32,595  WARN |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MembershipManager] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Member [127.0.0.1]:5702 - d594f66e-5397-4d70-8030-741f2c444a27 is suspected to be dead for reason: Connection manager is stopped on Member [127.0.0.1]:5702 - d594f66e-5397-4d70-8030-741f2c444a27 this
05:34:32,595  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MembershipManager] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Starting mastership claim process...
05:34:32,595  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MembershipManager] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Local MembersView{version=6, members=[MemberInfo{address=[127.0.0.1]:5702, uuid=d594f66e-5397-4d70-8030-741f2c444a27, liteMember=false, memberListJoinVersion=2}, MemberInfo{address=[127.0.0.1]:5704, uuid=8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, liteMember=false, memberListJoinVersion=5}]} with suspected members: [Member [127.0.0.1]:5702 - d594f66e-5397-4d70-8030-741f2c444a27] and initial addresses to ask: []
05:34:32,595  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Shutting down node engine...
05:34:32,595  WARN |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [InternalPartitionService] hz.pensive_spence.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Ignoring received partition table, startup is not completed yet. Sender: [127.0.0.1]:5702
05:34:32,595 ERROR |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [PromotionCommitOperation] hz.pensive_spence.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Promotion of 50 partitions failed. . Promotion partition state stamp: 2538535148351109352, current partition state stamp: -3954767352384688262
05:34:32,595  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [PartitionStateManager] hz.recursing_bardeen.cached.thread-3 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Storing snapshot of partition assignments while removing UUID d594f66e-5397-4d70-8030-741f2c444a27
05:34:32,595  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [ClusterService] hz.recursing_bardeen.cached.thread-3 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] 

Members {size:1, ver:7} [
	Member [127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2 this
]

05:34:32,595  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MembershipManager] hz.recursing_bardeen.cached.thread-3 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Mastership is claimed with: MembersView{version=7, members=[MemberInfo{address=[127.0.0.1]:5704, uuid=8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, liteMember=false, memberListJoinVersion=5}]}
05:34:32,595  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [TransactionManagerService] hz.recursing_bardeen.cached.thread-2 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5702, UUID: d594f66e-5397-4d70-8030-741f2c444a27
05:34:32,595  WARN |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Invocation] hz.pensive_spence.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Future.complete(Object) on completed future. Request: Invocation{op=com.hazelcast.internal.partition.operation.PromotionCommitOperation{serviceName='hz:core:partitionService', identityHash=-996065916, partitionId=-1, replicaIndex=0, callId=-1143, invocationTime=1652679272594 (2022-05-16 05:34:32.594), waitTimeout=-1, callTimeout=60000, tenantControl=com.hazelcast.spi.impl.tenantcontrol.NoopTenantControl@0}, tryCount=2147483647, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1652679272594, firstInvocationTime='2022-05-16 05:34:32.594', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[127.0.0.1]:5702, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=null}, current value: ExceptionalResult{cause=com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!}, offered value: false
java.lang.Exception: null
	at com.hazelcast.spi.impl.AbstractInvocationFuture.warnIfSuspiciousDoubleCompletion(AbstractInvocationFuture.java:1253) [classes/:?]
	at com.hazelcast.spi.impl.AbstractInvocationFuture.complete0(AbstractInvocationFuture.java:1230) [classes/:?]
	at com.hazelcast.spi.impl.AbstractInvocationFuture.complete(AbstractInvocationFuture.java:1219) [classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.complete(Invocation.java:673) [classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.sendResponse(Invocation.java:236) [classes/:?]
	at com.hazelcast.spi.impl.operationservice.Operation.sendResponse(Operation.java:483) [classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:283) [classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:248) [classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:213) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:175) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:139) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.executeRun(OperationThread.java:123) [classes/:?]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102) [classes/:?]
05:34:32,596  WARN |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MigrationManager] hz.pensive_spence.migration - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Promotion commit failed for [MigrationInfo{uuid=b56fb116-cfa0-4cba-9cc3-80b15d6589eb, partitionId=94, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=95c31aa4-20c0-4512-bbde-02d93271a0e0, partitionId=96, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=76c6b089-030c-4893-861c-3f460617a971, partitionId=97, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=afb647ce-5c7c-4600-b858-074de57c6c51, partitionId=98, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=0aa8996b-e0ae-4278-b08b-13ea94d28a3f, partitionId=101, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=69499b26-3351-45ea-b43d-9138a7989d59, partitionId=103, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=7c63e28b-e46f-4c62-a096-d2a00aa54a71, partitionId=104, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=fd0d3437-9804-4630-8c22-eff6aa223743, partitionId=105, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=12f99d0f-dcf3-408b-90b0-d555af3c7d83, partitionId=106, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=8f211fee-24a0-43e6-be96-b2e85d4f5a20, partitionId=107, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=349387ca-87c4-46df-8152-bcff06f98f64, partitionId=109, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=52c3433e-267c-4271-ae1d-2831f1022662, partitionId=112, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=76eeaa49-31a7-4ede-a04b-1eea21ebe99e, partitionId=115, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=37653280-6c00-4bf8-90b9-90696b38c715, partitionId=119, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=3ea08366-f1c5-4506-b249-c1e582ad04ca, partitionId=120, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=12c30d08-2d09-4048-ba58-9a23a1798e27, partitionId=121, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=2066b52a-577e-4186-97e7-beda8ab902d1, partitionId=122, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=b9fed7bc-43ec-481a-8163-8743f60d38d3, partitionId=126, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=2f9c5325-1738-4c1b-afab-aab2f0308f7c, partitionId=127, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=e7c91fd3-f9c1-4d38-a056-075a69a491b1, partitionId=128, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=778dd926-6e98-4241-ba72-8c34b7c62000, partitionId=130, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=fca75d8d-dca6-4d29-86aa-07e10de86517, partitionId=132, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=c0acbb57-1bac-4a61-b82d-fcccd0d1bd7b, partitionId=133, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=8de20517-a27a-4d3f-aaee-519b17bb3aaf, partitionId=135, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=35d533ee-1497-454a-b80b-703febc81fdd, partitionId=138, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=faf25d75-bda2-427d-9c8b-ca5cc5b8ed7f, partitionId=139, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=ad1743c8-c659-4d8f-b83e-8302e625958f, partitionId=144, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=5a920c1d-7835-4839-844a-45219b844ddd, partitionId=148, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=2a22427e-8ec7-4a49-a769-5ef2dc53f1f7, partitionId=149, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=6e2827b5-36ba-43ed-a4ed-e321639b895c, partitionId=150, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=5a47e4f1-5971-422f-ad70-764ff9315ad5, partitionId=152, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=7a46f25c-3050-4ee7-84c2-2446904270f0, partitionId=153, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=54ac03bc-933e-4853-93fb-97889bdff05a, partitionId=155, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=deca814f-0608-480f-9414-66f8a452eaa7, partitionId=157, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=fd2b5a3b-0821-4a7d-8753-74365e6e6f23, partitionId=158, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=7, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=ac6c3f5b-9980-418a-8775-0f0b6c88d643, partitionId=159, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=7fdf50e0-00e5-4b2e-9b1d-c16d7c04f060, partitionId=163, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=66db912d-11d3-4a92-a5f5-aa3abb22b616, partitionId=164, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=a1937a6c-662d-4330-96d4-5c7e13dc0c17, partitionId=170, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}, MigrationInfo{uuid=a6e25088-252c-42ca-bd6e-943210c92147, partitionId=171, source=null, sourceCurrentReplicaIndex=-1, sourceNewReplicaIndex=-1, destination=[127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2, destinationCurrentReplicaIndex=1, destinationNewReplicaIndex=0, master=[127.0.0.1]:5702, initialPartitionVersion=8, partitionVersionIncrement=2, status=SUCCESS}] since destination [127.0.0.1]:5704 - 8c19428a-bf0a-4e65-bb3b-cd768e49d1c2 left the cluster
05:34:32,845  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [InternalPartitionService] hz.recursing_bardeen.migration - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Fetching partition tables from cluster to determine the most recent one... Local stamp: -3271556229799755952
05:34:32,845  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [InternalPartitionService] hz.recursing_bardeen.migration - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Most recent partition table is determined.
05:34:32,845  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [InternalPartitionService] hz.recursing_bardeen.migration - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Applying the most recent of partition state...
05:34:32,845  WARN |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [InternalPartitionService] hz.recursing_bardeen.migration - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Following unknown addresses are found in partition table sent from master[[127.0.0.1]:5704]. (Probably they have recently joined or left the cluster.) {
	[127.0.0.1]:5702 - d594f66e-5397-4d70-8030-741f2c444a27
	[127.0.0.1]:5703 - 2311a9e1-3881-4749-8314-1934e8cf0aa5
}
05:34:32,865  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [MigrationManager] hz.recursing_bardeen.migration - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Partition balance is ok, no need to repartition.
05:34:33,306  WARN |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [RaftNodeImpl$LeaderFailureDetectionTask(METADATA)] hz.recursing_bardeen.partition-operation.thread-0 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Current leader RaftEndpoint{uuid='2311a9e1-3881-4749-8314-1934e8cf0aa5'} is not reachable. Will start new election round...
05:34:33,306  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [RaftNode(METADATA)] hz.recursing_bardeen.partition-operation.thread-0 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] 

CP Group Members {groupId: METADATA(0), size:3, term:2, logIndex:56} [
	CPMember{uuid=d594f66e-5397-4d70-8030-741f2c444a27, address=[127.0.0.1]:5702}
	CPMember{uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5, address=[127.0.0.1]:5703}
	CPMember{uuid=e6275a2d-ffd3-4320-9381-a1dfecbd8adf, address=[127.0.0.1]:5704} - FOLLOWER this
]

05:34:33,306  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [PreVoteTask(METADATA)] hz.recursing_bardeen.partition-operation.thread-0 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Pre-vote started for next term: 3, last log index: 57, last log term: 2
05:34:33,306  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [RaftNode(METADATA)] hz.recursing_bardeen.partition-operation.thread-0 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] 

CP Group Members {groupId: METADATA(0), size:3, term:2, logIndex:56} [
	CPMember{uuid=d594f66e-5397-4d70-8030-741f2c444a27, address=[127.0.0.1]:5702}
	CPMember{uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5, address=[127.0.0.1]:5703}
	CPMember{uuid=e6275a2d-ffd3-4320-9381-a1dfecbd8adf, address=[127.0.0.1]:5704} - FOLLOWER this
]

05:34:35,584  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [PreVoteTask(METADATA)] hz.recursing_bardeen.partition-operation.thread-0 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Pre-vote started for next term: 3, last log index: 57, last log term: 2
05:34:35,584  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [RaftNode(METADATA)] hz.recursing_bardeen.partition-operation.thread-0 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] 

CP Group Members {groupId: METADATA(0), size:3, term:2, logIndex:56} [
	CPMember{uuid=d594f66e-5397-4d70-8030-741f2c444a27, address=[127.0.0.1]:5702}
	CPMember{uuid=2311a9e1-3881-4749-8314-1934e8cf0aa5, address=[127.0.0.1]:5703}
	CPMember{uuid=e6275a2d-ffd3-4320-9381-a1dfecbd8adf, address=[127.0.0.1]:5704} - FOLLOWER this
]

05:34:35,597  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [NodeExtension] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Destroying node NodeExtension.
05:34:35,597  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Hazelcast Shutdown is completed in 3003 ms.
05:34:35,597  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [LifecycleService] Thread-13485 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5702 is SHUTDOWN
05:34:35,597  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [LifecycleService] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5704 is SHUTTING_DOWN
05:34:35,597  WARN |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Terminating forcefully...
05:34:35,597  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Shutting down connection manager...
05:34:35,597  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Shutting down node engine...
05:34:35,599  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [NodeExtension] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Destroying node NodeExtension.
05:34:35,599  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] Hazelcast Shutdown is completed in 2 ms.
05:34:35,599  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [LifecycleService] Thread-13485 - [127.0.0.1]:5704 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5704 is SHUTDOWN
05:34:35,599  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [LifecycleService] Thread-13485 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
05:34:35,599  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [Node] Thread-13485 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Node is already shutting down... Waiting for shutdown process to complete...
05:34:35,599  INFO |when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot| - [LifecycleService] Thread-13485 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot(com.hazelcast.cp.internal.CPMemberAddRemoveTest): BuildInfo{version='5.2-SNAPSHOT', build='20220516', buildNumber=20220516, revision=504ac9f, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstallsSnapshot(com.hazelcast.cp.internal.CPMemberAddRemoveTest):'
05:32:15, accumulated pauses: 82 ms, max pause: 30 ms, pauses over 1000 ms: 0
05:32:20, accumulated pauses: 222 ms, max pause: 79 ms, pauses over 1000 ms: 0
05:32:25, accumulated pauses: 161 ms, max pause: 62 ms, pauses over 1000 ms: 0
05:32:30, accumulated pauses: 96 ms, max pause: 64 ms, pauses over 1000 ms: 0
05:32:35, accumulated pauses: 134 ms, max pause: 55 ms, pauses over 1000 ms: 0
05:32:40, accumulated pauses: 83 ms, max pause: 45 ms, pauses over 1000 ms: 0
05:32:45, accumulated pauses: 68 ms, max pause: 41 ms, pauses over 1000 ms: 0
05:32:50, accumulated pauses: 68 ms, max pause: 41 ms, pauses over 1000 ms: 0
05:32:55, accumulated pauses: 26 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:33:00, accumulated pauses: 105 ms, max pause: 50 ms, pauses over 1000 ms: 0
05:33:05, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:33:10, accumulated pauses: 46 ms, max pause: 20 ms, pauses over 1000 ms: 0
05:33:15, accumulated pauses: 26 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:33:20, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:33:25, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:33:30, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:33:35, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:33:40, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:33:45, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:33:50, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:33:55, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:34:00, accumulated pauses: 28 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:34:05, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:34:10, accumulated pauses: 36 ms, max pause: 9 ms, pauses over 1000 ms: 0
05:34:15, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:34:20, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:34:25, accumulated pauses: 27 ms, max pause: 0 ms, pauses over 1000 ms: 0
05:34:30, accumulated pauses: 70 ms, max pause: 43 ms, pauses over 1000 ms: 0
05:34:35, accumulated pauses: 3 ms, max pause: 0 ms, pauses over 1000 ms: 0
```

</details>

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/MembershipChangeSchedule.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/MembershipChangeSchedule.java
@@ -26,11 +26,13 @@ import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.HashSet;
+import java.util.LinkedHashSet;
 import java.util.List;
+import java.util.Set;
 
 import static com.hazelcast.internal.util.Preconditions.checkNotNull;
 import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableSet;
 
 /**
  * When there is a membership change in CP Subsystem,
@@ -163,11 +165,13 @@ public class MembershipChangeSchedule implements IdentifiedDataSerializable {
             groupId = in.readObject();
             membersCommitIndex = in.readLong();
             int len = in.readInt();
-            members = new HashSet<>(len);
+            // The initial incoming implementation for the 'members' variable in a constructor is LinkedHashSet
+            Set<RaftEndpoint> membersSet = new LinkedHashSet<>(len);
             for (int i = 0; i < len; i++) {
                 RaftEndpoint member = in.readObject();
-                members.add(member);
+                membersSet.add(member);
             }
+            members = unmodifiableSet(membersSet);
             memberToAdd = in.readObject();
             memberToRemove = in.readObject();
         }

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/MetadataRaftGroupManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/MetadataRaftGroupManager.java
@@ -222,16 +222,9 @@ public class MetadataRaftGroupManager implements SnapshotAwareService<MetadataRa
             logger.fine("Taking snapshot for commit-index: " + commitIndex);
         }
 
-        MetadataRaftGroupSnapshot snapshot = new MetadataRaftGroupSnapshot();
-        snapshot.setMembers(activeMembers);
-        snapshot.setMembersCommitIndex(activeMembersCommitIndex);
-        snapshot.setGroups(groups.values());
-        snapshot.setMembershipChangeSchedule(membershipChangeSchedule);
-        snapshot.setInitialCPMembers(initialCPMembers);
-        snapshot.setInitializedCPMembers(initializedCPMembers);
-        snapshot.setInitializationStatus(initializationStatus);
-        snapshot.setInitializationCommitIndices(initializationCommitIndices);
-
+        MetadataRaftGroupSnapshot snapshot = new MetadataRaftGroupSnapshot(activeMembers, activeMembersCommitIndex,
+                groups.values(), membershipChangeSchedule, initialCPMembers, initializedCPMembers, initializationStatus,
+                initializationCommitIndices);
         return snapshot;
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/cp/internal/MetadataRaftGroupSnapshot.java
+++ b/hazelcast/src/main/java/com/hazelcast/cp/internal/MetadataRaftGroupSnapshot.java
@@ -27,9 +27,15 @@ import java.util.Collection;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.unmodifiableCollection;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableSet;
 
 /**
  * Snapshot of the Metadata Raft group state
+ * This class should be IMMUTABLE since it is stored in the RaftLog's SnapshotEntry
  */
 public final class MetadataRaftGroupSnapshot implements IdentifiedDataSerializable {
 
@@ -38,75 +44,63 @@ public final class MetadataRaftGroupSnapshot implements IdentifiedDataSerializab
     private final Collection<CPGroupInfo> groups = new ArrayList<>();
     private MembershipChangeSchedule membershipChangeSchedule;
     private List<CPMemberInfo> initialCPMembers;
-    private Set<CPMemberInfo> initializedCPMembers = new HashSet<>();
+    private final Set<CPMemberInfo> initializedCPMembers = new HashSet<>();
     private MetadataRaftGroupInitStatus initializationStatus;
-    private Set<Long> initializationCommitIndices = new HashSet<>();
+    private final Set<Long> initializationCommitIndices = new HashSet<>();
 
-    public void setGroups(Collection<CPGroupInfo> groups) {
-        for (CPGroupInfo group : groups) {
-            // Deep copy CPGroupInfo, because it's a mutable object.
-            this.groups.add(new CPGroupInfo(group));
-        }
+    public MetadataRaftGroupSnapshot() {
     }
 
-    public void setMembers(Collection<CPMemberInfo> members) {
+    public MetadataRaftGroupSnapshot(Collection<CPMemberInfo> members,
+                                     long membersCommitIndex,
+                                     Collection<CPGroupInfo> groups,
+                                     MembershipChangeSchedule membershipChangeSchedule,
+                                     List<CPMemberInfo> initialCPMembers,
+                                     Set<CPMemberInfo> initializedCPMembers,
+                                     MetadataRaftGroupInitStatus initializationStatus,
+                                     Set<Long> initializationCommitIndices) {
         this.members.addAll(members);
+        this.membersCommitIndex = membersCommitIndex;
+        // Deep copy CPGroupInfo, because it's a mutable object.
+        groups.stream().map(CPGroupInfo::new).forEach(this.groups::add);
+        this.membershipChangeSchedule = membershipChangeSchedule;
+        this.initialCPMembers = initialCPMembers;
+        this.initializedCPMembers.addAll(initializedCPMembers);
+        this.initializationStatus = initializationStatus;
+        this.initializationCommitIndices.addAll(initializationCommitIndices);
     }
 
     public Collection<CPMemberInfo> getMembers() {
-        return members;
+        return unmodifiableCollection(members);
     }
 
     public long getMembersCommitIndex() {
         return membersCommitIndex;
     }
 
-    public void setMembersCommitIndex(long membersCommitIndex) {
-        this.membersCommitIndex = membersCommitIndex;
-    }
-
     public Collection<CPGroupInfo> getGroups() {
-        return groups;
+        // Deep copy CPGroupInfo, because it's a mutable object.
+        return groups.stream().map(CPGroupInfo::new).collect(Collectors.toList());
     }
 
     public MembershipChangeSchedule getMembershipChangeSchedule() {
         return membershipChangeSchedule;
     }
 
-    public void setMembershipChangeSchedule(MembershipChangeSchedule membershipChangeSchedule) {
-        this.membershipChangeSchedule = membershipChangeSchedule;
-    }
-
     public Set<CPMemberInfo> getInitializedCPMembers() {
-        return initializedCPMembers;
-    }
-
-    public void setInitializedCPMembers(Collection<CPMemberInfo> initializedCPMembers) {
-        this.initializedCPMembers.addAll(initializedCPMembers);
+        return unmodifiableSet(initializedCPMembers);
     }
 
     public List<CPMemberInfo> getInitialCPMembers() {
-        return initialCPMembers;
-    }
-
-    public void setInitialCPMembers(List<CPMemberInfo> initialCPMembers) {
-        this.initialCPMembers = initialCPMembers;
+        return unmodifiableList(initialCPMembers);
     }
 
     public MetadataRaftGroupInitStatus getInitializationStatus() {
         return initializationStatus;
     }
 
-    public void setInitializationStatus(MetadataRaftGroupInitStatus initializationStatus) {
-        this.initializationStatus = initializationStatus;
-    }
-
     public Set<Long> getInitializationCommitIndices() {
-        return initializationCommitIndices;
-    }
-
-    public void setInitializationCommitIndices(Set<Long> initializationCommitIndices) {
-        this.initializationCommitIndices.addAll(initializationCommitIndices);
+        return unmodifiableSet(initializationCommitIndices);
     }
 
     @Override
@@ -187,5 +181,19 @@ public final class MetadataRaftGroupSnapshot implements IdentifiedDataSerializab
             long commitIndex = in.readLong();
             initializationCommitIndices.add(commitIndex);
         }
+    }
+
+    @Override
+    public String toString() {
+        return "MetadataRaftGroupSnapshot{"
+                + "members=" + members
+                + ", membersCommitIndex=" + membersCommitIndex
+                + ", groups=" + groups
+                + ", membershipChangeSchedule=" + membershipChangeSchedule
+                + ", initialCPMembers=" + initialCPMembers
+                + ", initializedCPMembers=" + initializedCPMembers
+                + ", initializationStatus=" + initializationStatus
+                + ", initializationCommitIndices=" + initializationCommitIndices
+                + '}';
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/cp/internal/CPMemberAddRemoveTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cp/internal/CPMemberAddRemoveTest.java
@@ -28,6 +28,7 @@ import com.hazelcast.cp.IAtomicLong;
 import com.hazelcast.cp.exception.CPGroupDestroyedException;
 import com.hazelcast.cp.internal.raft.impl.RaftNodeImpl;
 import com.hazelcast.cp.internal.raft.impl.command.UpdateRaftGroupMembersCmd;
+import com.hazelcast.cp.internal.raft.impl.log.SnapshotEntry;
 import com.hazelcast.cp.internal.raftop.metadata.GetActiveCPMembersOp;
 import com.hazelcast.cp.internal.raftop.metadata.GetMembershipChangeScheduleOp;
 import com.hazelcast.cp.internal.raftop.metadata.GetRaftGroupOp;
@@ -815,6 +816,57 @@ public class CPMemberAddRemoveTest extends HazelcastRaftTestSupport {
         });
     }
 
+    // test for https://github.com/hazelcast/hazelcast/issues/21438
+    @Test
+    public void when_snapshotIsTakenWhileRemovingCPLeader_newMemberInstalledSnapshot_shouldBeImmutable() throws Exception {
+        int nodeCount = 3;
+        int commitIndexAdvanceCountToSnapshot = 50;
+        Config config = createConfig(nodeCount, nodeCount);
+        config.getCPSubsystemConfig().getRaftAlgorithmConfig().setCommitIndexAdvanceCountToSnapshot(commitIndexAdvanceCountToSnapshot);
+
+        HazelcastInstance[] instances = new HazelcastInstance[nodeCount];
+        for (int i = 0; i < nodeCount; i++) {
+            instances[i] = factory.newHazelcastInstance(config);
+        }
+
+        assertClusterSizeEventually(nodeCount, instances);
+        waitUntilCPDiscoveryCompleted(instances);
+
+        HazelcastInstance leaderInstance = getLeaderInstance(instances, getMetadataGroupId(instances[0]));
+
+        // `commitIndexAdvanceCountToSnapshot - 5` is selected on purpose to partially include removal of CP member in snapshot.
+        // Specifically, RemoveCPMemberOp will be in snapshot but CompleteRaftGroupMembershipChangesOp will not.
+        for (int i = 0; i < commitIndexAdvanceCountToSnapshot - 5; i++) {
+            getRaftInvocationManager(leaderInstance).invoke(getMetadataGroupId(instances[0]), new GetActiveCPMembersOp()).get();
+        }
+
+        // This will add 3 entries, RemoveCPMemberOp, ChangeRaftGroupMembersCmd and CompleteRaftGroupMembershipChangesOp.
+        // RemoveCPMemberOp will be in snapshot but CompleteRaftGroupMembershipChangesOp will not be included.
+        leaderInstance.shutdown();
+
+        HazelcastInstance newInstance = factory.newHazelcastInstance(config);
+        newInstance.getCPSubsystem().getCPSubsystemManagementService().promoteToCPMember().toCompletableFuture().join();
+
+        List<CPMember> cpMembers = new ArrayList<>(newInstance.getCPSubsystem().getCPSubsystemManagementService().getCPMembers().toCompletableFuture().join());
+
+        // Waiting for newInstance to apply the snapshot and update its active group members
+        assertTrueEventually(() -> {
+            RaftService service = getRaftService(newInstance);
+            List<CPMemberInfo> activeMembers = new ArrayList<>(service.getMetadataGroupManager().getActiveMembers());
+            assertEquals(cpMembers, activeMembers);
+        });
+
+        // Getting SnapshotEntry from newInstance and leader instances RaftLog
+        RaftNodeImpl newInstanceRaftNode = getRaftNode(newInstance, getMetadataGroupId(newInstance));
+        SnapshotEntry newInstanceSnapshotEntry = (SnapshotEntry) getSnapshotEntry(newInstanceRaftNode);
+        RaftNodeImpl leaderRaftNode = getRaftNode(getInstance(newInstanceRaftNode.getLeader()), getMetadataGroupId(newInstance));
+        SnapshotEntry leaderSnapshotEntry = (SnapshotEntry) getSnapshotEntry(leaderRaftNode);
+
+        // Verifying that the newInstance's SnapshotEntry hasn't mutated after group members' state update,
+        // and equals to the initial leader's SnapshotEntry.
+        // The 'toString' method is used here to compare states so as not to add 'equals' methods to a bunch of classes.
+        assertEquals(leaderSnapshotEntry.toString(true), newInstanceSnapshotEntry.toString(true));
+    }
 
     @Test
     public void when_newCPMemberIsAddedToTheMetadataGroupAfterRestart_newMemberCommitsMetadataGroupLogEntries() throws ExecutionException, InterruptedException {

--- a/hazelcast/src/test/resources/log4j2.xml
+++ b/hazelcast/src/test/resources/log4j2.xml
@@ -38,9 +38,7 @@
         <Logger name="com.hazelcast.client.impl.spi.ClientListenerService" level="trace"/>
         <Logger name="com.hazelcast.jet" level="debug"/>
         <Logger name="com.hazelcast.sql.impl.client" level="debug"/>
-        <Logger name="com.hazelcast.cp.internal.raft.impl.RaftNodeImpl" level="debug"/>
-        <Logger name="com.hazelcast.cp.internal.MetadataRaftGroupManager" level="debug"/>
-        <Logger name="com.hazelcast.cp.internal.RaftGroupMembershipManager" level="debug"/>
+        <!--<Logger name="com.hazelcast.cp.internal.raft.impl" level="debug"/>-->
 <!--        <Logger name="com.hazelcast.internal.partition.operation.MigrationRequestOperation" level="trace"/>-->
         <!--<Logger name="com.hazelcast.client.impl.spi.ClientInvocationService" level="trace"/>-->
         <!--<Logger name="com.hazelcast.client.impl.connection.ClientConnectionManager" level="trace"/>-->

